qcodes/dataset/data_set_protocol.py

```diff
@@ -0,0 +1,158 @@
+SPECS = List[ParamSpec]
+# Transition period type: SpecsOrInterDeps. We will allow both as input to
+# the DataSet constructor for a while, then deprecate SPECS and finally remove
+# the ParamSpec class
+SpecsOrInterDeps = Union[SPECS, InterDependencies_]
+
+
+if TYPE_CHECKING:
+    from qcodes.station import Station
+
+    from .data_set_cache import DataSetCache
+
+
+@runtime_checkable
+class DataSetProtocol(Protocol, Sized):
+    def prepare(
+        self,
+        *,
+        station: Optional[Station],
+        interdeps: InterDependencies_,
+        write_in_background: bool,
+        shapes: Shapes = None,
+        parent_datasets: Sequence[Mapping[Any, Any]] = (),
+    ) -> None:
+        pass
+
+    def _enqueue_results(self, result_dict: Mapping[ParamSpecBase, np.ndarray]) -> None:
+        pass
+
+    def get_metadata(self, tag: str) -> str:
+        pass
+
+    def add_metadata(self, tag: str, metadata: Any) -> None:
+        pass
+
+    @property
+    def run_id(self) -> int:
+        pass
+
+    @property
+    def captured_run_id(self) -> int:
+        pass
+
+    @property
+    def captured_counter(self) -> int:
+        pass
+
+    @property
+    def guid(self) -> str:
+        pass
+
+    def add_snapshot(self, snapshot: str, overwrite: bool = False) -> None:
+        pass
+
+    def mark_completed(self) -> None:
+        pass
+
+    @property
+    def parent_dataset_links(self) -> List[Link]:
+        pass
+
+    def export(
+        self,
+        export_type: Optional[Union[DataExportType, str]] = None,
+        path: Optional[str] = None,
+        prefix: Optional[str] = None,
+    ) -> None:
+        pass
+
+    @property
+    def number_of_results(self) -> int:
+        pass
+
+    def _flush_data_to_database(self, block: bool = False) -> None:
+        pass
+
+    @property
+    def paramspecs(self) -> Dict[str, ParamSpec]:
+        pass
+
+    @property
+    def exp_name(self) -> str:
+        pass
+
+    @property
+    def sample_name(self) -> str:
+        pass
+
+    @property
+    def name(self) -> str:
+        pass
+
+    @property
+    def snapshot_raw(self) -> Optional[str]:
+        pass
+
+    @property
+    def run_timestamp_raw(self) -> Optional[float]:
+        pass
+
+    def run_timestamp(self, fmt: str = "%Y-%m-%d %H:%M:%S") -> Optional[str]:
+        pass
+
+    @property
+    def completed_timestamp_raw(self) -> Optional[float]:
+        pass
+
+    def completed_timestamp(self, fmt: str = "%Y-%m-%d %H:%M:%S") -> Optional[str]:
+        pass
+
+    @property
+    def completed(self) -> bool:
+        pass
+
+    @completed.setter
+    def completed(self, value: bool) -> None:
+        pass
+
+    @property
+    def description(self) -> RunDescriber:
+        pass
+
+    @property
+    def metadata(self) -> Dict[str, Any]:
+        pass
+
+    @property
+    def dependent_parameters(self) -> Tuple[ParamSpecBase, ...]:
+        pass
+
+    @property
+    def cache(self) -> DataSetCache[DataSetProtocol]:
```

Comment:
get_parameter_data is not part of the protocol, which means that in type-checked code whenever ``datasaver.dataset`` is used to call ``get_parameter_data`` mypy will be confused (right?). what is the way to avoid this? switch to using ``dataset.cache.data()``? and if the specifics of ``get_parameter_data`` are required, then users would use ``cast``?

Reply:
None of the accessors are part of the interface and yes I do indeed want to move all of this to the cache

Reply:
ok. but the `cast` part will probably be annoying, but let's see